File: src/main/scala/PurpleAir.scala
package smick

import com.twitter.conversions.DurationOps._
import com.twitter.finagle.Http
import com.twitter.finagle.http.RequestBuilder
import com.twitter.util.Future
import java.net.URL

case class PurpleAirData(
  mapVersion: String,
  baseVersion: String,
  mapVersionString: String,
  results: List[Map[String, Any]])

trait PurpleAir { self: SmickHome =>

  val purpleAirFreq = flag("purpleair.freq", 10.seconds, "Purple Air polling frequency")
  val purpleAirSensor = flag("purpleair.sensor", 0, "Purple Air sensor ID")

  def purpleAirLoop(store: Store): Future[Unit] =
    loopIt("purpleAir", purpleAirFreq(), process(store))

  private[this] val url = Lazy[URL] {
    new URL(s"https://www.purpleair.com/json?show=${purpleAirSensor()}")
  }

  private[this] val client = Lazy[HttpSvc] {
    val name = "purple-air"
    Http.client
      .withTls(url().getHost)
      .newService(destName(name, url()), name)
  }

  private[this] val sensorMetrics =
    Seq(
      "p_0_3_um",
      "p_0_5_um",
      "p_1_0_um",
      "p_2_5_um",
      "p_5_0_um",
      "p_10_0_um",
      "pm1_0_cf_1",
      "pm2_5_cf_1",
      "pm10_0_cf_1",
      "pm1_0_atm",
      "pm2_5_atm",
      "pm10_0_atm",
      "humidity",
      "temp_f",
      "pressure")

  private def process(store: Store): Future[Unit] =
    client()(RequestBuilder().url(url()).buildGet()) flatMap { res =>
      val data = json.readValue[PurpleAirData](res.contentString)

      val entries = for {
        result <- data.results
        tags = Map("label" -> result.get("Label").get)
        metric <- sensorMetrics
        v <- result.get(metric)
      } yield StoreEntry(metric, v, tags)

      store.write(entries)
    }
}

File: src/internal/looper/purpleair.go
package looper

import (
	"context"
	"fmt"
	"github.com/rs/zerolog"
	"github.com/spf13/pflag"
	hm "github.com/sprsquish/housemetrics/internal"
	"github.com/sprsquish/housemetrics/internal/store"
	"net/url"
	"strconv"
	"time"
)

type PurpleAir struct {
	client *hm.HttpClient
	logger *zerolog.Logger

	sensor int
	url    *url.URL
}

type PurpleAirReading struct {
	Results []map[string]interface{}
}

var metricNames = []string{
	"p_0_3_um",
	"p_0_5_um",
	"p_1_0_um",
	"p_2_5_um",
	"p_5_0_um",
	"p_10_0_um",
	"pm1_0_cf_1",
	"pm2_5_cf_1",
	"pm10_0_cf_1",
	"pm1_0_atm",
	"pm2_5_atm",
	"pm10_0_atm",
	"humidity",
	"temp_f",
	"pressure",
}

func NewPurpleAir(name string, flags *pflag.FlagSet, logger *zerolog.Logger, client *hm.HttpClient) hm.Looper {
	p := PurpleAir{
		client: client,
		logger: logger,
	}

	flags.IntVar(&p.sensor, fmt.Sprintf("%s.sensor", name), 0, "Sensor ID")

	return &p
}

func (p *PurpleAir) Init() {
	p.logger.Debug().Msg("inititalizing")
	sensorURL, err := url.Parse(fmt.Sprintf("https://www.purpleair.com/json?show=%d", p.sensor))
	if err != nil {
		p.logger.Error().Err(err).Msg("couldn't parse URL")
		return
	}
	p.logger.Debug().Interface("url", sensorURL).Msg("setting url")
	p.url = sensorURL
}

func (p *PurpleAir) Poll(ctx context.Context, store store.Client) error {
	var reading PurpleAirReading
	if err := p.client.GetJSON(ctx, p.logger, &reading, hm.URLOpt(p.url)); err != nil {
		return err
	}

	tags := map[string]string{"label": ""}
	ts := time.Now()
	for _, result := range reading.Results {
		tags["label"] = result["Label"].(string)
		for _, name := range metricNames {
			var val interface{}

			switch v := result[name].(type) {
			case string:
				val, _ = strconv.ParseFloat(v, 16)
			default:
				val = v
			}

			if val != nil {
				store.Write(ctx, ts, name, val, tags)
			}
		}
	}

	return nil
}
